src/main/scala/mesosphere/marathon/core/group/impl/AssignDynamicServiceLogic.scala

```diff
@@ -0,0 +1,108 @@
+    *
+    * Assignment logic is:
+    * 1) First, scan for service ports used in old app, but are not specified in the new app. Use these first.
+    * 2) Second, consume from provided unassignedPortsIterator
+    *
+    * If no ports are assigned, there are no side effects.
+    *
+    * @return The updated app definition with assigned service ports
+    */
+  private def assignPorts(newApp: AppDefinition, oldApp: Option[AppDefinition], portRange: Range,
+    unassignedPortsIterator: Iterator[Int]): AppDefinition = {
+    /* All ports that are already assigned in old app definition, but not used in the new definition
+     * if the app uses dynamic ports (0), it will get always the same ports assigned */
+    val assignedAndAvailable: Seq[Int] =
+      oldApp match {
+        case Some(oldApp) =>
+          oldApp.servicePorts.filter { p: Int => portRange.contains(p) && !newApp.servicePorts.contains(p) }
+        case None => Nil
+      }
+
+    val freePortsIterator = assignedAndAvailable.iterator ++ unassignedPortsIterator
+
+    val servicePorts: Seq[Int] = newApp.servicePorts.map {
+      case 0 =>
+        if (freePortsIterator.hasNext)
+          freePortsIterator.next()
+        else
+          throw new PortRangeExhaustedException(portRange.min, portRange.max)
+      case port => port
+    }
+
+    val updatedContainer = newApp.container.find(_.portMappings.nonEmpty).map { container =>
+      val newMappings = container.portMappings.zip(servicePorts).map {
+        case (portMapping, servicePort) => portMapping.copy(servicePort = servicePort)
+      }
+      container.copyWith(portMappings = newMappings)
+    }
+
+    newApp.copy(
+      portDefinitions = mergeServicePortsAndPortDefinitions(newApp.portDefinitions, servicePorts),
+      container = updatedContainer.orElse(newApp.container))
+  }
+
+  def assignDynamicServicePorts(portRange: Range, from: RootGroup, to: RootGroup): RootGroup = {
+    /* Note: We consider the "from" rootGroup servicePorts as well, since these are reserved for re-use for a specific
+     * app in the case that servicePorts are re-posted all as 0's.
+     */
+    val usedServicePorts: Set[Int] =
+      (from.transitiveApps.iterator ++ to.transitiveApps.iterator).flatMap(_.servicePorts).toSet
+    val unassignedPortsIterator = portRange.iterator
+      .filter { p => !usedServicePorts.contains(p) }
+      .map { port =>
+        logger.debug(s"Take next configured free port: $port")
+        port
+      }
+    val dynamicApps: Iterator[AppDefinition] =
+      to.transitiveApps
+        .iterator
+        .filter { newApp => changedOrNew(from.transitiveAppsById, newApp) }
+        .map {
+          // assign values for service ports that the user has left "blank" (set to zero)
+          case app: AppDefinition if app.hasDynamicServicePorts =>
+            assignPorts(app, from.app(app.id), portRange, unassignedPortsIterator)
+          case app: AppDefinition =>
```

Comment:
this is hurting readability a bit because it looks like mergeServicePortsAndPortDefinitions is only done for app that does not have dynamic service ports - but that is not true, because it is also hidden in the assignPorts before returning the value...

But I guess doing it like this we save ourselves one more extra copy so we will probably have to live with this..